File: rest-service/src/main/java/mb/lib/blast/model/p/BlastpTask.java
package mb.lib.blast.model.p;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import mb.api.service.util.ErrorText;

public enum BlastpTask
{
  BlastP("blastp"),
  BlastPFast("blastp-fast"),
  BlastPShort("blastp-short");

  private final String value;

  BlastpTask(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return value;
  }

  @JsonCreator
  public static BlastpTask fromString(String value) {
    for (var e : values())
      if (e.value.equals(value))
        return e;

    throw new IllegalArgumentException(String.format(
      ErrorText.InvalidEnumValue,
      value,
      BlastpTask.class.getSimpleName()
    ));
  }
}

File: common/blast-types/src/main/kotlin/org/veupathdb/lib/blast/blastp/field/BlastPTask.kt
package org.veupathdb.lib.blast.blastp.field

import com.fasterxml.jackson.databind.node.ObjectNode
import org.veupathdb.lib.blast.common.FlagDust
import org.veupathdb.lib.blast.common.FlagTask
import org.veupathdb.lib.blast.serial.BlastField
import org.veupathdb.lib.blast.util.*


internal fun ParseBlastPTask(js: ObjectNode) =
  js.optString(FlagTask) { BlastPTask(parseTask(it)) } ?: BlastPTask()


/**
 * -task `<String>`
 *
 * Permissible values:
 * * blastp
 * * blastp-fast
 * * blastp-short
 *
 * Task to execute
 *
 * Default = `blastp`
 */
@JvmInline
value class BlastPTask(val value: BlastPTaskType = BlastPTaskType.BlastP)
  : BlastField
{
  override val isDefault get() = value == BlastPTaskType.BlastP

  override val name: String
    get() = FlagTask

  override fun appendJson(js: ObjectNode) =
    js.put(isDefault, FlagTask, value.value)

  override fun appendCliSegment(cli: StringBuilder) =
    cli.append(isDefault, FlagTask, value.value)

  override fun appendCliParts(cli: MutableList<String>) =
    cli.add(isDefault, FlagTask, value.value)
}


private fun parseTask(value: String) = when(value) {
  "blastp"       -> BlastPTaskType.BlastP
  "blastp-fast"  -> BlastPTaskType.BlastPFast
  "blastp-short" -> BlastPTaskType.BlastPShort
  else           -> throw IllegalArgumentException("Invalid value $value for enum BlastPTaskType")
}


enum class BlastPTaskType {
  BlastP,
  BlastPFast,
  BlastPShort;

  val value get() = when(this) {
    BlastP      -> "blastp"
    BlastPFast  -> "blastp-fast"
    BlastPShort -> "blastp-short"
  }
}